src/mlpack/core/tree/spill_tree/single_tree_traverser.hpp

```diff
@@ -0,0 +1,64 @@
+/**
+ * @file single_tree_traverser.hpp
+ * @author Ryan Curtin
+ * @author Marcos Pividori
+ *
+ * A nested class of SpillTree which traverses the entire tree with a
+ * given set of rules which indicate the branches which can be pruned and the
+ * order in which to recurse.  This traverser is a depth-first traverser.
+ */
+#ifndef MLPACK_CORE_TREE_SPILL_TREE_SINGLE_TREE_TRAVERSER_HPP
+#define MLPACK_CORE_TREE_SPILL_TREE_SINGLE_TREE_TRAVERSER_HPP
+
+#include <mlpack/core.hpp>
+
+#include "spill_tree.hpp"
+
+namespace mlpack {
+namespace tree {
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+template<typename RuleType>
+class SpillTree<MetricType, StatisticType, MatType, HyperplaneType, SplitType>::
+    SingleTreeTraverser
```

Comment:
Hi @rcurtin,
Thanks for your comments, I agree!
I have modified the SpillTree's Traversers. Now they take a template boolean parameter `Defeatist`, that determines if the traverser must consider defeatist search on overlapping nodes. In the commits:
https://github.com/mlpack/mlpack/pull/747/commits/0267acb46caef51370e737cf253a1664ea7be263
https://github.com/mlpack/mlpack/pull/747/commits/70fbeab1e0be29113f0e21c688601133c23d4751
https://github.com/mlpack/mlpack/pull/747/commits/b334674a12817b3e4b3f37ac6f99eb52bfd3e44c

So, for `SpillTrees`, I provide 4 tree traversers:
+`SingleTreeTraverser`   implements a normal single tree traverser.
+`DualTreeTraverser`   implements a normal dual tree traverser.
+`DefeatistSingleTreeTraverser`   does defeatist search on overlapping nodes.
+`DefeatistDualTreeTraverser`   does defeatist serach on overlapping node.